playground/backend/internal/code_processing/code_processing.go

```diff
@@ -210,30 +154,122 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
+		var compileOutput bytes.Buffer
+		runCmdWithOutput(compileCmd, &compileOutput, &compileError, successChannel, errorChannel)
+
+		// Start of the monitoring of background tasks (compile step/cancellation/timeout)
+		ok, err := reconcileBackgroundTask(pipelineLifeCycleCtx, ctx, pipelineId, cacheService, cancelChannel, successChannel)
+		if err != nil {
+			return nil
+		}
+		if !ok { // Compile step is finished, but code couldn't be compiled (some typos for example)
+			err := <-errorChannel
+			_ = processErrorWithSavingOutput(pipelineLifeCycleCtx, err, compileError.Bytes(), pipelineId, cache.CompileOutput, cacheService, "Compile", pb.Status_STATUS_COMPILE_ERROR)
+			return nil
+		} // Compile step is finished and code is compiled
+		if err := processCompileSuccess(pipelineLifeCycleCtx, compileOutput.Bytes(), pipelineId, cacheService); err != nil {
+			return nil
+		}
+	}
+	return &executor
+}
+
+func prepareStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) *executors.Executor {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupPreparatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return nil
+	}
+	executor := executorBuilder.Build()
+	// Prepare
+	logger.Infof("%s: Prepare() ...\n", pipelineId)
+	prepareFunc := executor.Prepare()
+	// Run prepare function
+	go prepareFunc(successChannel, errorChannel, validationResults)
+
+	// Start of the monitoring of background tasks (prepare function/cancellation/timeout)
+	ok, err := reconcileBackgroundTask(pipelineLifeCycleCtx, ctx, pipelineId, cacheService, cancelChannel, successChannel)
+	if err != nil {
+		return nil
+	}
+	if !ok {
+		err := <-errorChannel
+		// Prepare step is finished, but code couldn't be prepared (some error during prepare step)
+		_ = processErrorWithSavingOutput(pipelineLifeCycleCtx, err, []byte(err.Error()), pipelineId, cache.PreparationOutput, cacheService, "Prepare", pb.Status_STATUS_PREPARATION_ERROR)
+		return nil
+	}
+	// Prepare step is finished and code is prepared
+	if err := processSuccess(pipelineLifeCycleCtx, pipelineId, cacheService, "Prepare", pb.Status_STATUS_COMPILING); err != nil {
+		return nil
+	}
+	return &executor
+}
+
+func validateStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) *executors.Executor {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupValidatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return nil
+	}
+	executor := executorBuilder.Build()
+	// Validate
```

Comment:
This comment is "extremely useful", let's remove it😁

Reply:
Done